File: python/47.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def backtrack(start):
            if start == n:
                res.append(nums[:])
            used = set()
            for i in range(start, n):
                if nums[i] in used:
                    continue
                used.add(nums[i])
                nums[i], nums[start] = nums[start], nums[i]
                backtrack(start + 1)
                nums[i], nums[start] = nums[start], nums[i]
        
        n = len(nums)
        res = []
        backtrack(0)
        return res

File: C#/47.cs
public class Solution {
    public IList<IList<int>> PermuteUnique(int[] nums) {
        IList<IList<int>> result = new List<IList<int>>();
        Array.Sort(nums);
        Backtrack(result, new List<int>(), nums, new bool[nums.Length]);
        return result;
    }
    
    private void Backtrack(IList<IList<int>> result, List<int> tempList, int[] nums, bool[] used) {
        if(tempList.Count == nums.Length) {
            result.Add(new List<int>(tempList));
        } else {
            for(int i = 0; i < nums.Length; i++) {
                if(used[i] || (i > 0 && nums[i] == nums[i-1] && !used[i-1])) continue;
                used[i] = true;
                tempList.Add(nums[i]);
                Backtrack(result, tempList, nums, used);
                used[i] = false;
                tempList.RemoveAt(tempList.Count - 1);
            }
        }
    }
}